arangod/Cluster/ClusterInfo.cpp

```diff
@@ -3724,6 +3722,78 @@ std::shared_ptr<std::vector<ServerID>> ClusterInfo::getResponsibleServer(ShardID
 
   return std::make_shared<std::vector<ServerID>>();
 }
+  
+//////////////////////////////////////////////////////////////////////////////
+/// @brief atomically find all servers who are responsible for the given 
+/// shards (only the leaders).
+/// will throw an exception if no leader can be found for any
+/// of the shards. will return an empty result if the shards couldn't be
+/// determined after a while - it is the responsibility of the caller to
+/// check for an empty result!
+//////////////////////////////////////////////////////////////////////////////
+
+std::unordered_map<ShardID, ServerID> ClusterInfo::getResponsibleServers(std::unordered_set<ShardID> const& shardIds) {
+  TRI_ASSERT(!shardIds.empty());
+
+  std::unordered_map<ShardID, ServerID>  result;
+  int tries = 0;
+
+  if (!_currentProt.isValid) {
+    loadCurrent();
+    tries++;
+  }
+
+  while (true) {
+    TRI_ASSERT(result.empty());
+    {
+      READ_LOCKER(readLocker, _currentProt.lock);
+      for (auto const& shardId : shardIds) {
+        auto it = _shardIds.find(shardId);
+
+        if (it == _shardIds.end()) {
+          THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND, "no servers found for shard " + shardId);
+        }
+
+        auto serverList = (*it).second;
+        if (serverList == nullptr || serverList->empty()) {
+          THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "no servers found for shard " + shardId);
+        }
+
+        if ((*serverList)[0].size() > 0 && (*serverList)[0][0] == '_') {
+          // This is a temporary situation in which the leader has already
+          // resigned, let's wait half a second and try again.
+          --tries;
+          break;
+        }
+
+        // put leader into result
+        result.emplace(shardId, (*it).second->front());
+      }
+    }
+
+    if (result.size() == shardIds.size()) {
+      // result is complete
+      break;
+    }
+    
+    // reset everything we found so far for the next round
+    result.clear();
+    
+    if (++tries >= int(2 * shardIds.size()) || application_features::ApplicationServer::isStopping()) {
```

Comment:
I think we can actually keep this at 3 instead of magic formula.
AFAIK this is only to cover against ongoing move jobs, those are done in parallel and should report in reasonable amount of time.
Also if we wait here for let's say 3 collections 9 shards each => `54 * 0.5s == 27s` all other parts of the query might be rather unhappy anyways...